test/e2e/clusterclass_rollout.go

```diff
@@ -75,6 +75,9 @@ type ClusterClassRolloutSpecInput struct {
 	// Allows to inject a function to be run after test namespace is created.
 	// If not specified, this is a no-op.
 	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
+
+	// ProviderSpecificLabels represents a set of labels specific to the provider.
+	ProviderSpecificLabels map[string]string
```

Comment:
I was wondering if we should add a func like this
```
ValidateObjectMetadata func(obj client.Object) (clusterv1.ObjectMeta, error)
```

The idea is that folks can set this func and then validate metadata + return an updated version of ObjectMetadata (labels/annotaitons) that can then pass the core CAPI validation.

Basically the CAPV func would check:
* if the object is a Machine, does the Machine have the right label
  * If no => return an error
  * If yes => return a modified version of ObjectMeta without the additional label

This allows us a lot of flexibility for all label & annotation validation. (including that we don't have to specifcy correct label values up-front)

Reply:
@chrischdi @fabriziopandini WDYT?

Reply:
@zhanggbj Talked about it with Christian and Fabrizio.

We would prefer adding a field like this:
```go
	// FilterMetadataBeforeValidation allows filtering out labels and annotations of Machines, InfraMachines,
	// BootstrapConfigs and Nodes before we validate them.
	// This can be e.g. used to filter out additional infrastructure provider specific labels that would
	// otherwise lead to a failed test.
	FilterMetadataBeforeValidation func(object client.Object) clusterv1.ObjectMeta
```

This would then be roughly used like this

```go
				machineMetadata := clusterv1.ObjectMeta{Labels: machine.Labels, Annotations: machine.Annotations}
				if input.FilterMetadataBeforeValidation != nil {
					machineMetadata = input.FilterMetadataBeforeValidation(machine)
				}

				// MachineDeployment MachineSet Machine.metadata
				g.Expect(machineMetadata.Labels).To(BeEquivalentTo(
					union(
						map[string]string{
							clusterv1.ClusterNameLabel:                          cluster.Name,
							clusterv1.ClusterTopologyOwnedLabel:                 "",
							clusterv1.ClusterTopologyMachineDeploymentNameLabel: mdTopology.Name,
							clusterv1.MachineDeploymentNameLabel:                machineDeployment.Name,
							clusterv1.MachineSetNameLabel:                       machineSet.Name,
							clusterv1.MachineDeploymentUniqueLabel:              machineTemplateHash,
						},
						machineSet.Spec.Template.Labels,
					),
				))
				g.Expect(machineMetadata.Annotations).To(BeEquivalentTo(
					machineSet.Spec.Template.Annotations,
				))
```

Reply:
Hi @sbueringer, thanks for providing the suggestions!

I just updated assertControlPlaneMachines and assertMachineSetsMachines to ensure I understand the concept correctly. If there are no issues, I'll proceed with updating the other objects in https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/clusterclass_rollout.go#L311-L323.  Thanks!

Reply:
Overall look goods (please see comments below). You already got all objects covered that we would use this func for. Currently we only think we have use cases for these